crates/gitbutler-branch-actions/src/branch_manager/branch_removal.rs

```diff
@@ -0,0 +1,248 @@
+        };
+
+        // We don't want to try unapplying branches which are marked as not in workspace by the new metric
+        if !branch.in_workspace {
+            return Ok(());
+        }
+
+        _ = self
+            .project_repository
+            .project()
+            .snapshot_branch_deletion(branch.name.clone());
+
+        let repo = self.project_repository.repo();
+
+        let integration_commit = repo.integration_commit()?;
+        let target_commit = repo.target_commit()?;
+        let base_tree = target_commit.tree().context("failed to get target tree")?;
+
+        let virtual_branches = vb_state
+            .list_branches_in_workspace()
+            .context("failed to read virtual branches")?;
+
+        let (applied_statuses, _) = get_applied_status(
+            self.project_repository,
+            &integration_commit.id(),
+            virtual_branches,
+        )
+        .context("failed to get status by branch")?;
+
+        // go through the other applied branches and merge them into the final tree
+        // then check that out into the working directory
+        let final_tree = applied_statuses
+            .into_iter()
+            .filter(|(branch, _)| branch.id != branch_id)
+            .fold(
+                target_commit.tree().context("failed to get target tree"),
+                |final_tree, status| {
+                    let final_tree = final_tree?;
+                    let branch = status.0;
+                    let tree_oid = write_tree(self.project_repository, &branch.head, status.1)?;
+                    let branch_tree = repo.find_tree(tree_oid)?;
+                    let mut result =
+                        repo.merge_trees(&base_tree, &final_tree, &branch_tree, None)?;
+                    let final_tree_oid = result.write_tree_to(repo)?;
+                    repo.find_tree(final_tree_oid)
+                        .context("failed to find tree")
+                },
+            )?;
+
+        // checkout final_tree into the working directory
+        repo.checkout_tree_builder(&final_tree)
+            .force()
+            .remove_untracked()
+            .checkout()
+            .context("failed to checkout tree")?;
+
+        vb_state
+            .mark_as_not_in_workspace(branch.id)
+            .context("Failed to remove branch")?;
+
+        self.project_repository.delete_branch_reference(&branch)?;
+
+        ensure_selected_for_changes(&vb_state).context("failed to ensure selected for changes")?;
+
+        Ok(())
+    }
+}
+
+trait PrivateBranchRemoval {
+    fn build_metadata_commit(
+        &self,
+        vbranch: &mut branch::Branch,
+        branch: &git2::Branch<'_>,
+    ) -> Result<git2::Oid>;
+    fn build_real_branch(
+        &self,
+        vbranch: &mut branch::Branch,
+        name_conflict_resolution: NameConflitResolution,
+    ) -> Result<git2::Branch<'_>>;
+}
+
+impl PrivateBranchRemoval for BranchManager<'_> {
+    fn build_real_branch(
```

Comment:
it's used only in one place above. consider making this a simple function. or feel free to try this as is 🤷